src/cirrus/lib2/events.py

```diff
@@ -0,0 +1,249 @@
+
+        if payload is not None and payload["id"] != payload_id:
+            raise ValueError(
+                "payload_id and payload['id'] must match if both supplied."
+            )
+        message = {
+            "event_type": event_type,
+            "payload": payload,
+            "payload_id": payload_id,
+        }
+
+        if extra_message:
+            if not all(k not in message for k in extra_message):
+                raise ValueError(
+                    "extra_message parameters must not include: "
+                    + f"{','.join(message.keys())}."
+                )
+            message.update(extra_message)
+        self.event_publisher.add(
+            message=json.dumps(message),
+            message_attrs={
+                "event_type": {"DataType": "String", "StringValue": event_type}
+            },
+        )
+
+    def claim_processing(self, payload: dict, isotimestamp: str = None):
+        if not isotimestamp:
+            isotimestamp = datetime.now(timezone.utc).isoformat()
+        self.statedb.claim_processing(
+            payload_id=payload["id"], isotimestamp=isotimestamp
+        )
+        self.announce(WFEventType.CLAIMED_PROCESSING, payload=payload)
+
+    def started_processing(
+        self, payload: dict, execution_arn: str, isotimestamp: str = None
+    ):
+        if not isotimestamp:
+            isotimestamp = datetime.now(timezone.utc).isoformat()
+        self.statedb.set_processing(payload["id"], execution_arn, isotimestamp)
+        self.write_timeseries_record(
+            payload["id"],
+            state=StateEnum.PROCESSING,
+            event_time=isotimestamp,
+            execution_arn=execution_arn,
+        )
+        self.announce(
+            WFEventType.STARTED_PROCESSING,
+            payload=payload,
+            extra_message={"execution_arn": execution_arn},
+        )
+
+    def skipping(self, payload: dict, state: StateEnum):
+        self.logger.warning(f"already in {state} state: {payload['id']}")
+        self.announce(WFEventType(f"ALREADY_{state}"), payload)
+
+    def duplicated(self, payload: dict):
+        self.logger.warning("duplicate payload_id dropped %s", payload["id"])
+        self.announce(WFEventType.DUPLICATE_ID_ENCOUNTERED, payload)
+
+    def failed(
+        self,
+        payload: dict,
+        message: str = "",
+        execution_arn=None,
+        isotimestamp: str = None,
+    ):
+        if not isotimestamp:
+            isotimestamp = datetime.now(timezone.utc).isoformat()
+        self.statedb.set_failed(payload["id"], message, execution_arn=execution_arn)
+        if execution_arn:
+            self.write_timeseries_record(
+                payload["id"], StateEnum.FAILED, isotimestamp, execution_arn
+            )
+
+        self.announce(
+            WFEventType.FAILED,
+            payload,
+            extra_message={"error": message, "execution_arn": execution_arn},
+        )
+
+    def timed_out(
```

Comment:
Is this not a failure with regard to the timestream events as well?

Reply:
Yes.  And added.